### PR TITLE
Fix flaky tests in WindowsServiceLifetimeTests

### LINKED ISSUE #93194: Test failure Microsoft.Extensions.Hosting.WindowsServiceLifetimeTests.ServiceSequenceIsCorrect
**Failed in:** [runtime-coreclr libraries-pgo 20231008.1](https://dev.azure.com/dnceng-public/public/_build/results?buildId=431320&view=ms.vss-test-web.build-test-results-tab&runId=9538040&resultId=109275&paneView=debug)

**Failed tests:**
```
net9.0-windows-Release-x86-CoreCLR_checked-syntheticpgo_blend-Windows.10.Amd64.Open
    - Microsoft.Extensions.Hosting.WindowsServiceLifetimeTests.ServiceSequenceIsCorrect
```

**Error message:**
```
Assert.Equal() Failure
                                   ↓ (pos 74)
Expected: ···Service.StartAsync\r\nlifetime started\r\nWindowsServiceLifetime.···
Actual:   ···Service.StartAsync\r\nWindowsServiceLifetime.OnStop\r\nlifetime s···
                                   ↑ (pos 74)
```

**Stack trace:**
```
   at Microsoft.Extensions.Hosting.WindowsServiceLifetimeTests.ServiceSequenceIsCorrect() in /_/src/libraries/Microsoft.Extensions.Hosting.WindowsServices/tests/WindowsServiceLifetimeTests.cs:line 242
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
   at System.Reflection.MethodBaseInvoker.InvokeWithNoArgs(Object obj, BindingFlags invokeAttr) in /_/src/libraries/System.Private.CoreLib/src/System/Reflection/MethodBaseInvoker.cs:line 48
```

### LINKED ISSUE #103262: Test failure: Microsoft.Extensions.Hosting.WindowsServiceLifetimeTests.ServiceCanStopItself
**Failed in:** [runtime-coreclr libraries-pgo 20240609.1](https://dev.azure.com/dnceng-public/public/_build/results?buildId=701806&view=ms.vss-test-web.build-test-results-tab&runId=17471672&resultId=176621&paneView=debug)

**Failed tests:**
```
net9.0-windows-Release-x86-fullpgo_random_gdv-Windows.10.Amd64.Open
    - Microsoft.Extensions.Hosting.WindowsServiceLifetimeTests.ServiceCanStopItself
```

**Error message:**
```
 System.ServiceProcess.TimeoutException : The operation requested for service 'ServiceCanStopItself' has not been completed within the specified time interval.
```

**Stack trace:**
```
   at System.ServiceProcess.ServiceController.WaitForStatus(ServiceControllerStatus desiredStatus, TimeSpan timeout) in /_/src/libraries/System.ServiceProcess.ServiceController/src/System/ServiceProcess/ServiceController.cs:line 994
   at Microsoft.Extensions.Hosting.WindowsServiceTester.WaitForStatus(ServiceControllerStatus desiredStatus) in /_/src/libraries/Microsoft.Extensions.Hosting.WindowsServices/tests/WindowsServiceTester.cs:line 54
   at Microsoft.Extensions.Hosting.WindowsServiceLifetimeTests.ServiceCanStopItself() in /_/src/libraries/Microsoft.Extensions.Hosting.WindowsServices/tests/WindowsServiceLifetimeTests.cs:line 174
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
   at System.Reflection.MethodBaseInvoker.InvokeWithNoArgs(Object obj, BindingFlags invokeAttr) in /_/src/libraries/System.Private.CoreLib/src/System/Reflection/MethodBaseInvoker.cs:line 48
```

### PR DESCRIPTION
**Pull request by buyaa-n**
- `ServiceCanStopItself` test fails with `System.ServiceProcess.TimeoutException : The operation requested for service 'ServiceCanStopItself' has not been completed within the specified time interval.` The interval is 30 milliseconds, changing it to `TimeSpan.MaxValue` because the test is expected to stop eventually, it's not defined what time interval will be enough.
- `ServiceSequenceIsCorrect` test fails because of race condition, it looks `serviceTester.Stop()` started before `serviceTester.Start()` completed. It was suggested to `introduce a synchronization primitive here and have the test wait for that to be signaled by the service`. Added `AutoResetEvent` instance that signals when windows service started completely

Fixes https://github.com/dotnet/runtime/issues/103262
Fixes https://github.com/dotnet/runtime/issues/93194

### CODE REVIEW

File: src/libraries/Microsoft.Extensions.Hosting.WindowsServices/tests/WindowsServiceLifetimeTests.cs [hunk 1]
```diff
@@ -194,6 +198,8 @@ lifetime stopped
                 """, logText);
         }
 
+        private static AutoResetEvent are = new AutoResetEvent(false);
```
Comment (steveharter) [inline src/libraries/Microsoft.Extensions.Hosting.WindowsServices/tests/WindowsServiceLifetimeTests.cs:201]: nit: use `s_are`

Comment (steveharter) [reply src/libraries/Microsoft.Extensions.Hosting.WindowsServices/tests/WindowsServiceLifetimeTests.cs:201]: Use 2 different events?

Comment (buyaa-n) [reply src/libraries/Microsoft.Extensions.Hosting.WindowsServices/tests/WindowsServiceLifetimeTests.cs]: > nit: use s_are

Done

> Use 2 different events?

Updated the windows service `OnStop()` to wait only for the `ServiceSequenceIsCorrect` test

**Review by ericstj — APPROVED** [review]
Looks reasonable to me.  Make sure you run the impacted tests in case they don't run in normal PR validation.

**Comment by steveharter** [comment]
> Looks reasonable to me. Make sure you run the impacted tests in case they don't run in normal PR validation.

Ran a few times locally in admin command prompt + Windows:
```
  Discovering: Microsoft.Extensions.Hosting.WindowsServices.Tests (app domain = on [no shadow copy], method display = ClassAndMethod, method display options = None)
  Discovered:  Microsoft.Extensions.Hosting.WindowsServices.Tests (found 11 of 12 test cases)
  Starting:    Microsoft.Extensions.Hosting.WindowsServices.Tests (parallel test collections = on [20 threads], stop on fail = off)
   Microsoft.Extensions.Hosting.WindowsServiceLifetimeTests.ServiceStops [STARTING]
    Microsoft.Extensions.Hosting.UseWindowsServiceTests.DefaultsToOffOutsideOfService [STARTING]
    Microsoft.Extensions.Hosting.UseWindowsServiceTests.DefaultsToOffOutsideOfService [FINISHED] Time: 0.1712331s
    Microsoft.Extensions.Hosting.UseWindowsServiceTests.ServiceCollectionExtensionMethodAddsWindowsServiceLifetimeInsideOfService [STARTING]
    Microsoft.Extensions.Hosting.UseWindowsServiceTests.ServiceCollectionExtensionMethodAddsWindowsServiceLifetimeInsideOfService [FINISHED] Time: 0.0623185s
    Microsoft.Extensions.Hosting.UseWindowsServiceTests.ServiceCollectionExtensionMethodDefaultsToOffOutsideOfService [STARTING]
    Microsoft.Extensions.Hosting.UseWindowsServiceTests.ServiceCollectionExtensionMethodDefaultsToOffOutsideOfService [FINISHED] Time: 0.0086278s
    Microsoft.Extensions.Hosting.UseWindowsServiceTests.CanCreateService [STARTING]
    Microsoft.Extensions.Hosting.WindowsServiceLifetimeTests.ServiceStops [FINISHED] Time: 0.4462791s
    Microsoft.Extensions.Hosting.WindowsServiceLifetimeTests.ExceptionOnStopIsPropagated [STARTING]
    Microsoft.Extensions.Hosting.UseWindowsServiceTests.CanCreateService [FINISHED] Time: 0.4668568s
    Microsoft.Extensions.Hosting.UseWindowsServiceTests.ServiceCollectionExtensionMethodCanBeCalledOnDefaultConfiguration [STARTING]
    Microsoft.Extensions.Hosting.UseWindowsServiceTests.ServiceCollectionExtensionMethodCanBeCalledOnDefaultConfiguration [FINISHED] Time: 0.0022696s
    Microsoft.Extensions.Hosting.UseWindowsServiceTests.ServiceCollectionExtensionMethodSetsEventLogSourceNameToApplicationNameInsideOfService [STARTING]
    Microsoft.Extensions.Hosting.UseWindowsServiceTests.ServiceCollectionExtensionMethodSetsEventLogSourceNameToApplicationNameInsideOfService [FINISHED] Time: 0.0015367s
    Microsoft.Extensions.Hosting.WindowsServiceLifetimeTests.ExceptionOnStopIsPropagated [FINISHED] Time: 0.3454064s
    Microsoft.Extensions.Hosting.WindowsServiceLifetimeTests.ServiceSequenceIsCorrect [STARTING]
    Microsoft.Extensions.Hosting.WindowsServiceLifetimeTests.ServiceSequenceIsCorrect [FINISHED] Time: 1.4618799s
    Microsoft.Extensions.Hosting.WindowsServiceLifetimeTests.CancelStopAsync [STARTING]
    Microsoft.Extensions.Hosting.WindowsServiceLifetimeTests.CancelStopAsync [FINISHED] Time: 0.0821135s
    Microsoft.Extensions.Hosting.WindowsServiceLifetimeTests.ServiceCanStopItself [STARTING]
    Microsoft.Extensions.Hosting.WindowsServiceLifetimeTests.ServiceCanStopItself [FINISHED] Time: 0.4354667s
  Finished:    Microsoft.Extensions.Hosting.WindowsServices.Tests
=== TEST EXECUTION SUMMARY ===
   Microsoft.Extensions.Hosting.WindowsServices.Tests  Total: 11, Errors: 0, Failed: 0, Skipped: 0, Time: 2.840s
```

**Comment by steveharter** [comment]
Merging; Buyaa out

**Pull request merged**